### ISSUE TITLE
'assertion failed: x <= i32::max_value() as f64'

### ISSUE THREAD

**Issue by nab-os** [WanzenBug/rqrr#15]
Hi,

I am experiencing crashes on a long running app.
Here is the error:
`thread 'Compute Task Pool (3)' panicked at 'assertion failed: x <= i32::max_value() as f64', /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/rqrr-0.5.1/src/geometry.rs:47:9`
```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/panicking.rs:48:5
   3: rqrr::geometry::Perspective::map
   4: rqrr::detect::create_capstone
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/rqrr-0.5.1/src/detect.rs:232:18
   5: rqrr::detect::capstones_from_image
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/rqrr-0.5.1/src/detect.rs:48:31
   6: rqrr::prepare::PreparedImage<S>::detect_grids
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/rqrr-0.5.1/src/prepare.rs:214:22
   7: mariage_vivien_marine::scan_user_id
             at ./src/main.rs:84:17
   8: core::ops::function::FnMut::call_mut
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/ops/function.rs:164:5
   9: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/ops/function.rs:290:13
  10: <Func as bevy_ecs::system::function_system::SystemParamFunction<(),Out,(F0,),()>>::run::call_inner
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.8.1/src/system/function_system.rs:564:21
  11: <Func as bevy_ecs::system::function_system::SystemParamFunction<(),Out,(F0,),()>>::run
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.8.1/src/system/function_system.rs:567:17
  12: <bevy_ecs::system::function_system::FunctionSystem<In,Out,Param,Marker,F> as bevy_ecs::system::system::System>::run_unsafe
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.8.1/src/system/function_system.rs:403:19
  13: bevy_ecs::schedule::executor_parallel::ParallelExecutor::prepare_systems::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.8.1/src/schedule/executor_parallel.rs:194:30
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/future/mod.rs:91:19
  15: async_executor::Executor::spawn::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.4.1/src/lib.rs:144:19
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/future/mod.rs:91:19
  17: async_task::raw::RawTask<F,T,S>::run
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.3.0/src/raw.rs:511:20
  18: async_executor::Executor::run::{{closure}}::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.4.1/src/lib.rs:235:21
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/future/mod.rs:91:19
  20: <futures_lite::future::Or<F1,F2> as core::future::future::Future>::poll
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:529:33
  21: async_executor::Executor::run::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.4.1/src/lib.rs:242:31
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/core/src/future/mod.rs:91:19
  23: futures_lite::future::block_on::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:89:27
  24: std::thread::local::LocalKey<T>::try_with
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/std/src/thread/local.rs:445:16
  25: std::thread::local::LocalKey<T>::with
             at /rustc/2287107588d92889d282e6cd3c1ca5df34cd34a5/library/std/src/thread/local.rs:421:9
  26: futures_lite::future::block_on
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:79:5
  27: bevy_tasks::task_pool::TaskPool::new_internal::{{closure}}::{{closure}}
             at /home/nabos/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.8.1/src/task_pool.rs:120:25
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The qrcode decoding is plugged to a v4l stream source and the decoding is run inside a thread on the Bevy game engine.
I need this app to be able to run all day.
Il you need anything to help debug let me know.

**Comment by WanzenBug** [comment]
If you could provide a sample image that triggers that panic, that would be great.

Currently I suspect something is wrong with the perspective calculation, as this seems to get triggered by calculating the "middle" of a recognized Capstone, but I'm not sure how that could produce values outside the i32 range, as presumably the corners of the capstone are all in i32 range

**Comment by nab-os** [comment]
Well, it doesn't seem to depend on what my camera captures.
I left the app running multiple hours without touching anything and at one point it crashes.
And I tried that several times with my camera pointing in different directions, with different scenes in front of it.

Still, I will try to save the image on disk before the decoding at each frame to retrieve the last image seen and load it back in manually to see if it comes from this.

I'll keep you updated

**Comment by nab-os** [comment]
I've tested my protocol and against what I thought, that is indeed the last image that crashes the app.
I loaded it manually and it crashed instantly.
Here is a beautiful capture from my desk:
![image](https://user-images.githubusercontent.com/9805355/193649742-89c047fc-5666-43a9-8ded-1f0702308aca.png)

Let me know what you find with this or if you need any more information on my setup/code.

**Comment by WanzenBug** [comment]
Thank you, with that image I could reproduce the issue. I've pushed a fix in 0.5.2.

Please let me know if that fixes the panics, there might be more causes still hidden. I guess running it 24/7 against some cam image is good fuzzing set-up 🙃

**Comment by nab-os** [comment]
It fixed the issue, it's almost a day since I applied the patch and it seem to be working well.
Thank you for your work and your efficiency !

I'll let you know if anything happens :)
